dbus-wrapper/dbus_client.c

```diff
@@ -0,0 +1,328 @@
+	if (r < 0) {
+		ERR("Can't read request results: %s", strerror(-r));
+		abort();
+	}
+
+	r = sd_event_exit(event, code);
+	if (r < 0) {
+		ERR("Can't exit event loop: %s", strerror(-r));
+		abort();
+	}
+
+	return 0;
+}
+
+static int on_run_command(sd_event_source *s, void *userdata)
+{
+	command_ctx_t *ctx = userdata;
+	struct list *opts = list_head(ctx->opts);
+	sd_bus_error error = SD_BUS_ERROR_NULL;
+	sd_bus_message *reply = NULL;
+	sd_bus_message *m = NULL;
+	int return_value = 0;
+	int r;
+
+	r = sd_bus_message_new_method_call(ctx->bus, &m,
+					   "org.O1.swupdd.Client",
+					   "/org/O1/swupdd/Client",
+					   "org.O1.swupdd.Client",
+					   ctx->method);
+	if (r < 0) {
+		ERR("Failed to create method call: %s", strerror(-r));
+		goto finish;
+	}
+	r = sd_bus_message_open_container(m, SD_BUS_TYPE_ARRAY, "{sv}");
+	while (opts) {
+		command_option_t *option = opts->data;
+		r = sd_bus_message_open_container(m, SD_BUS_TYPE_DICT_ENTRY, "sv");
+		r = sd_bus_message_append_basic(m, 's', option->name);
+		switch (option->type) {
+		case TYPE_STRING:
+			r = sd_bus_message_open_container(m, SD_BUS_TYPE_VARIANT, "s");
+			r = sd_bus_message_append_basic(m, 's', option->value.as_str);
+			break;
+		case TYPE_BOOL:
+			r = sd_bus_message_open_container(m, SD_BUS_TYPE_VARIANT, "b");
+			r = sd_bus_message_append_basic(m, 'b', &option->value.as_bool);
+			break;
+		case TYPE_INT:
+			r = sd_bus_message_open_container(m, SD_BUS_TYPE_VARIANT, "i");
+			r = sd_bus_message_append_basic(m, 'i', &option->value.as_int);
+			break;
+		default:
+			ERR("Unknown type");
+			abort();
+		}
+		r = sd_bus_message_close_container(m); /* SD_BUS_TYPE_VARIANT */
+		if (r < 0) {
+			ERR("Failed to close variant container: %s", strerror(-r));
+			goto finish;
+		}
+		r = sd_bus_message_close_container(m); /* SD_BUS_TYPE_DICT_ENTRY */
+		if (r < 0) {
+			ERR("Failed to close dict entry container: %s", strerror(-r));
+			goto finish;
+		}
+		opts = opts->next;
+	}
+	r = sd_bus_message_close_container(m); /* SD_BUS_TYPE_ARRAY */
+	if (r < 0) {
+		ERR("Failed to close array container: %s", strerror(-r));
+		goto finish;
+	}
+
+	if (ctx->argv_type == DBUS_CMD_SINGLE_ARG) {
+		r = sd_bus_message_append_basic(m, 's', ctx->argv[0] ? ctx->argv[0] : "");
+		if (r < 0) {
+			ERR("Failed to append argument: %s", strerror(-r));
+			goto finish;
+		}
+	} else if (ctx->argv_type == DBUS_CMD_MULTIPLE_ARGS) {
+		r = sd_bus_message_append_strv(m, ctx->argv);
+		if (r < 0) {
+			ERR("Failed to append arguments: %s", strerror(-r));
+			goto finish;
+		}
+	}
+
+	r = sd_bus_call(ctx->bus, m, 0, &error, &reply);
+	if (r < 0) {
+		ERR("Failed to make D-Bus call: %s (%s)",
+		    strerror(sd_bus_error_get_errno(&error)),
+		    error.message);
+		goto finish;
+	}
+
+	/* TODO: do we really need return_value? */
+	r = sd_bus_message_read(reply, "b", &return_value);
+	if (r < 0) {
+		ERR("Failed to parse response message: %s", strerror(-r));
+		goto finish;
+	}
+
+	if (!return_value) {
```

Comment:
I'm a little confused here, the return value corresponds to the dbus return value? So the read was able to work fine but the dbus call had some sort of issue indicated by the return_value?